### ISSUE TITLE
cmap 4 idDelta is a signed short

### ISSUE THREAD

**Issue by simoncozens** [fonttools/fonttools#2575]
I just got sent a font file which fails MS fontval validation due to problems with the cmap table. It's a CJK font and has lots of glyphs scattered around lots of codepoints.  Debugging this, I noticed one immediate problem:

https://docs.microsoft.com/en-us/typography/opentype/spec/cmap#format-4-segment-mapping-to-delta-values : 
```
int16 | idDelta[segCount] | Delta for all character codes in segment.
```

https://docs.python.org/3/library/struct.html:

Format | C Type | Python type | Standard size | Notes
-- | -- | -- | -- | --
h | short | integer | 2 | (2)
H | unsigned short | integer | 2 | (2)

https://github.com/fonttools/fonttools/blob/cace698bb029d58cdae00603140438fe4993bfa6/Lib/fontTools/ttLib/tables/_c_m_a_p.py#L907

That should be "h", right?

**Comment by anthrotype** [comment]
yes! 'h' is for signed, 'H' for unsigned, so `int16` is 'h'. Good catch! PR?

**Comment by simoncozens** [comment]
Wait a second, I think there is something more clever going on here. I think those delta array are being computed as twos-complement values. But I am still seeing cmap tables with start code - id delta < 0. Debugging some more.

**Comment by anthrotype** [comment]
indeed. cmap4 is basically the default table for most fonts (within Unicode BMP) so i'd be surprised if fonttools produced broken cmap4 tables..

**Comment by justvanrossum** [comment]
The Apple specification lists it as `UInt16`, so MS seems to have changed this.

https://docs.microsoft.com/en-us/typography/opentype/spec/cmap

Both specs say "The idDelta arithmetic is modulo 65536". Not sure how this affects the issue.

**Comment by justvanrossum** [comment]
I meant to link to https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6cmap.html, duh.

**Comment by simoncozens** [comment]
This is definitely a broken cmap table:

```
* FAIL MS-FonVal: The idRangeOffset is 0 and idDelta + startCode is negative DETAILS:
   * PlatID = 0, EncID = 3, Fmt = 4, idDelta[182] = -32329, startCode[182] = 12330
   * PlatID = 0, EncID = 3, Fmt = 4, idDelta[190] = -32573, startCode[190] = 12586
   * PlatID = 3, EncID = 1, Fmt = 4, idDelta[182] = -32329, startCode[182] = 12330
   * PlatID = 3, EncID = 1, Fmt = 4, idDelta[190] = -32573, startCode[190] = 12586 [code: fontval-error]
```
   
I know cmap4 is the main way we produce fonts, but I imagine there are corner cases which would only be tickled by fonts which come close to the 64k limit and jump around a lot between glyph ID and Unicode codepoint.

**Comment by justvanrossum** [comment]
```python
>>> import struct
>>> i = -100
>>> ui = struct.unpack("H", struct.pack("h", i))[0]
>>> x = 200
>>> (x + i) % 0x10000
100
>>> (x + ui) % 0x10000
100
```

**Comment by anthrotype** [comment]
so "the arithmetic is modulo 65536" means it doesn't matter signed or unsigned and fontval is being picky?

**Comment by justvanrossum** [comment]
Wild guess: the font you're looking at would not build if `idDeltaArray` were "h"...

**Comment by simoncozens** [comment]
Correct - I tried that.

**Comment by PeterCon** [comment]
FWIW, an [issue](https://github.com/MicrosoftDocs/typography-issues/issues/980) related to this was opened on the OT spec due to discussion and (IIUC) disagreement in an [issue](https://github.com/HinTak/Font-Validator/issues/72) in HinTak's FontVal repo. That led me to investigate this further.

MS and Adobe did make a change from unsigned to signed at some point in drafting the first version of the OT spec (it was USHORT in a v0.2 draft, but changed to SHORT by v1.01). I haven't discovered why; perhaps it was because of the example that follows (which originated in the earliest version of Apple's TrueType spec) with idDelta values < 0 suggested that signed was actually intended. In fact, unsigned was _not_ intended (and MS's rasterizer has always used USHORT), though the example was a bit misleading. 

If interested, I've added more comments in the above issues, including some details I learned on the C/C++ language specifications regarding behaviours of "overflow" for arithmetic operations (and why the spec can't really be disambiguated simply by providing a C snippet to indicate the intent).
